src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/Completeness/SqlSaveProductCompletenesses.php

```diff
@@ -118,4 +136,57 @@ private function channelIdsIndexedByChannelCodes(): array
 
         return $result;
     }
+
+    /**
+     * To avoid to get several dead lock exceptions in a row, we sleep between the retry. It lets the database take a breath and finish the other concurrent transactions triggering the deadlock.
+     * There is a random sleep as well, to avoid to restart at the same time the other concurrent processes doing a retry as well.
+     */
+    private function executeWithRetry(callable $function): void
+    {
+        $retry = 0;
+        $isError = true;
+        while (true === $isError) {
+            try {
+                $this->connection->transactional($function);
+
+                $isError = false;
+            } catch (DeadlockException $e) {
+                $retry += 1;
+
+                if (5 === $retry) {
+                    throw $e;
+                }
+
+                $this->logger->warning(sprintf('Deadlock occurred when persisting the completeness, %s/4 retry', $retry));
+                usleep(300000 + rand(50000, $retry * 100000));
+            }
+        }
+    }
+
+    /**
+     * We don't catch any exception if an error occurs, because it's the last attempt to insert the data by locking the
+     * completeness table.
+     * Do note that LOCK TABLE locks also the table in READ mode for all the foreign keys (locale, channel and product tables).
+     * It means that a concurrent transaction can't insert data in the product table at the same time (just read). That's why the foreign check constraint is deactivated to avoid these locks.
+     */
+    private function executeWithLockOnTable(callable $function): void
+    {
+        $this->logger->warning('Locking the whole completeness table to persist the completeness, as it fails after trying 5 times to insert data due to deadlocks.');
+
+        $value = $this->connection->executeQuery('SELECT @@autocommit')->fetch();
+        if (!isset($value['@@autocommit']) && ((int) $value['@@autocommit'] !== 1 || (int) $value['@@autocommit'] !== 0)) {
```

Comment:
I only detect this now but I think we have a condition pb on master too ;)

```suggestion
        if (!isset($value['@@autocommit']) || ((int) $value['@@autocommit'] !== 1 || (int) $value['@@autocommit'] !== 0)) {
```
or
```suggestion
        if (isset($value['@@autocommit']) && ((int) $value['@@autocommit'] !== 1 || (int) $value['@@autocommit'] !== 0)) {
```
otherwise this exception will never be thrown and a notice will be fired :)

Reply:
Whaoo, thanks Aymeric, this is the code I copy from master... So it seems master doesn't work properly...